Re-point OS API calls to Codefi proxy

adonesky1 opened the pull request:
Routes all requests we make to OpenSea in `controllers` through the proxy server recently deployed by Codefi.

sethkfman commented:
This looks pretty straight forward. @andrepimenta @gantunesr  Any thoughts on the failover for the OpenSea proxy?

sethkfman commented on src/assets/CollectibleDetectionController.ts:
```diff
@@ -145,12 +145,20 @@ export class CollectibleDetectionController extends BaseController<
-        const collectiblesArray = await response.json();
+        try {
+          response = await timeoutFetch(api, {}, 15000);
+        } catch {
```
Would it be possible to throw this and then we can prompt the user before we decide to use the OpenSea API directly? If they decided they don't need the data I don't want to force an OpenSea API call without user consent.

adonesky1 replied:
We won't ever even hit this point unless the user has turned on the OpenSea API. I don't know if we need to surface a difference to the user between calling our proxy of OpenSea and OpenSea directly?

adonesky1 replied:
The controller won't detect at all while `disabled` is true. It is set to [true by default](https://github.com/MetaMask/controllers/blob/main/src/assets/CollectibleDetectionController.ts#L220) and only flip false if the user [has opted to turn on preferences for OpenSea API and NFT AutoDetection](https://github.com/MetaMask/controllers/blob/main/src/assets/CollectibleDetectionController.ts#L232)

Gudahtt commented on src/assets/CollectibleDetectionController.ts:
```diff
@@ -145,12 +145,20 @@ export class CollectibleDetectionController extends BaseController<
-        const collectiblesArray = await response.json();
+        try {
+          response = await timeoutFetch(api, {}, 15000);
+        } catch {
```
By omitting the `error` being caught here, we're suppressing all errors thrown in this `try` block. If something unexpected gets thrown, something indicative of a dev problem on our end (like invalid params, broken auth, malformed URL, etc.) we'd never know.

adonesky1 replied:
We are [currently swallowing errors in this same method](https://github.com/MetaMask/controllers/blob/main/src/assets/CollectibleDetectionController.ts#L159) but I agree we shouldn't be doing this.

adonesky1 replied:
ping here @Gudahtt

Gudahtt replied:
Generally we should let "programmer errors"/"unexpected errors" bubble up to be caught by Sentry if possible.

For errors that we expect to happen in some circumstances, like those caused by a misbehaving or offline server or a network issue, there is no one-size-fits-all approach. I'm not sure what would be most appropriate here. I didn't have anything specific in mind. It sounds like you've settled on logging for at least some of them.

adonesky1 replied:
I don't currently know of any errors we can deem "expected" or more of warnings that we can safely ignore. So I suppose we should log all errors until we become aware of some that might fit that description. In `controllers` how do we send errors to sentry?  The `safelyExecute` utility method that we wrap a bunch of stuff in seems to just `console.error` all errors. Do these get sent to sentry somehow?

Gudahtt replied:
This is a fetch call, so we should expect it to fail for any network error or server error or timeout, for starters.

Gudahtt replied:
There is no way to directly send errors to Sentry from this package aside from bubbling them up to the project. The `safetyExecute` utility should be considered deprecated; it swallows errors completely by default, and if enabled it logs them but only to the console, so we don't see them in prod.

adonesky1 replied:
> This is a fetch call, so we should expect it to fail for any network error or server error or timeout, for starters.

Sorry I don't think I follow what you think we should do on these types of errors. Are you saying if we get certain status codes we should swallow them?

adonesky1 replied:
And is [this a good example](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/controllers/metametrics.js#L86) of the current pattern/method we should be using to bubble errors from a controller to sentry (minus the exceptionsToFilter)? @Gudahtt

Gudahtt replied:
No, I didn't express any opinion on what to do about those errors. Those were examples of "expected errors", you said you couldn't think of any examples.

We need to ensure unexpected errors are logged to Sentry, so that we can catch our own bugs. Errors that we expect to happen, operation errors like network failures, should be treated on a case-by-case basis in whatever way we think is appropriate for the situation. As I said, I'm not sure what would be most appropriate here.

Gudahtt replied:
>And is [this a good example](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/controllers/metametrics.js#L86) of the current pattern/method we should be using to bubble errors from a controller to sentry (minus the exceptionsToFilter)

By "bubble up" what I mean is throw the exception, letting it proceed up the call stack until caught by a `catch` block or by Sentry's global unhandled error handler. That's generally what I'd recommend for any errors we don't anticipate.

Gudahtt replied:
See this guide for more error handling advice: https://www.joyent.com/node-js/production/design/errors

Gudahtt replied:
On the options for handling fetch errors, here are a few basic options:

* Swallow the error completely
  * Pros: no runtime cost
  * Cons: Hides potential problems
  If we _really, really_ don't care about the error, this might be appropriate. e.g. If we're using errors as control flow, like for validation, we don't need to log to console if `validateJson(badJson)` throws. But I can't think of an example where this makes sense for a fetch call. We always want a fetch call to succeed.
* Print the error to console
  * Pros: Highlights problem in development
  * Cons: Problem remains hidden in prod, and could flood console if it's too high-frequency.
  This is a good solution in cases where we _really_ don't care about the error, we just want to move on. The log is there for visibility. We can set it as debug level if it's a very mundane scenario that we usually don't care about, but usually warn or error level is more appropriate for a fetch error.
* Log the failure with a metric event
  * Pros: Lets us track frequency of failures in prod, increasing our confidence in the API/service or alerting us to prod failures, or failures that affect certain user segments
  * Cons: More complex to setup, and risks flooding Segment with events if frequency is too high.

Additional options that could be combined with the above strategies:
* We could try retrying the error (with or without logging it), maybe with an exponential backoff strategy to give the servers time to recover if they're down due to high traffic or a DDoS attack. We'd want to be careful that the error actually is something that might succeed upon retry though, or risk wasting time, bandwidth, and increasing server costs.
* We could limit error logging (to console or Segment) by time frame, such that we only log once per minute/hour/day or something like that, as a safeguard against flooding the console or Segment with events.
* We could use some sort of error sampling or aggregation technique. e.g. report 1st, 10th, 100th, and 1000th error, or report the number of errors per day, or report once per day if there has been "downtime" for a particular API.
* We could use browser APIs to detect internet connectivity, so that we don't even try if the user is offline. This would ensure any error logs/events we get are for more interesting problems. We don't care when the user is out of WiFi range, but we might care if the service tends to be blocked by corporate proxies or doesn't work with certain ISPs, etc.

adonesky1 replied:
I could see value in doing this to reduce the ever growing noise in our logs:
> We could use browser APIs to detect internet connectivity, so that we don't even try if the user is offline. This would ensure any error logs/events we get are for more interesting problems.

In terms of retrying, this method is already called within a polling cycle and the implications for failure are relatively low impact, we simply won't detect new NFTs that OpenSea has indexed to this account until the service is back up when we try again.

I guess my vote is to log fetch errors to the console since we will likely be alerted to these issues through codefi's monitoring of this proxy server which they are managing. Alternatively we could log them to segment so we could get some metrics about how frequently this impacting users?

adonesky1 replied:
Does this make sense to you @Gudahtt?

Gudahtt replied:
Yep! Either of those plans make sense to me

adonesky1 replied:
@Gudahtt here's an attempt to address some of your concerns: https://github.com/MetaMask/controllers/pull/805/commits/b21e4e5a7138e44b50cd79c37e1f7374831e6464

Regarding 
> We could use browser APIs to detect internet connectivity, so that we don't even try if the user is offline. 

I see [navigator.Online](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/onLine), what do think of pulling this API in as part of more widely used utility method to check before making any network requests?

adonesky1 replied:
ping here when you get a chance @Gudahtt

adonesky1 replied:
Totally know you're doing much more important stuff but going to keep pinging you just in case haha @Gudahtt

adonesky1 replied:
Another friendly ping here @Gudahtt

Gudahtt replied:
>I see [navigator.Online](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/onLine), what do think of pulling this API in as part of more widely used utility method to check before making any network requests?

Is that API supported by React Native, or a polyfill that we're already using? I wasn't sure how feasible using that is at this layer.

adonesky1 replied:
> Is that API supported by React Native, or a polyfill that we're already using?

Good call out. 
This API is available for the extension at this layer but is almost certainly not supported in React Native... Not seeing a polyfill or any alternative? Any ideas on this @gantunesr or @sethkfman?

adonesky1 commented on src/assets/CollectibleDetectionController.ts:
```diff
@@ -312,72 +316,71 @@ export class CollectibleDetectionController extends BaseController<
       return;
     }
 
-    await safelyExecute(async () => {
```
Not sure if we're prepared to actually remove this and have truly unexpected errors percolate beyond this? @Gudahtt

Gudahtt replied:
Nothing ever follows this call in this repository, so it's perfectly safe here. It is called in the products, so this change is probably worth calling out in the changelog. In the extension for example, it looks like there is a call in `actions.js` that could result in the loading screen getting stuck if this throws. That was already fallible though because of the connection... so that was always a risk, but now more so with this change.

I do think removing this is a good idea, if there are no cases you can anticipate where this would throw in production.

Gudahtt commented on src/assets/CollectiblesController.ts:
```diff
@@ -430,13 +444,36 @@ export class CollectiblesController extends BaseController<
+          },
+        );
+      } else {
+        return {
```
Nit: this might be easier to follow of this was assigned to `apiCollectibleContractObject`, leaving only one `return` statement. Or drop that variable, and return in all 3 places.

Gudahtt commented on src/assets/CollectiblesController.ts:
```diff
@@ -236,14 +237,27 @@ export class CollectiblesController extends BaseController<
+
+      if (this.openSeaApiKey) {
+        collectibleInformation = await handleFetch(
+          `https://api.opensea.io/api/v1/asset/${contractAddress}/${tokenId}`,
```
Nit: This could be moved into a `get___` function, to be more in-line with how we generate the `tokenURI`. Same with the `asset_contract` and `assets` APIs. Having a method for building each API call seems like a nice pattern.

Or it could be built into the existing function, with a flag to prefer OpenSea? :thinking: :shrug:

Gudahtt commented on src/assets/CollectiblesController.ts:
```diff
@@ -236,14 +237,27 @@ export class CollectiblesController extends BaseController<
+          name: null,
+          description: null,
+          image: null,
+          standard: null,
```
What is this property for? It's not returned on the other paths

adonesky1 replied:
returning out here allows us to skip that and the re-assigning step and just use the more easily encoded default/null version of the `CollectibleMetadata` which is ultimately return type of this function...

Gudahtt replied:
Yep! I must have missed it

Gudahtt commented on src/assets/CollectiblesController.ts:
```diff
@@ -236,14 +237,27 @@ export class CollectiblesController extends BaseController<
+          },
+        );
+      } else {
+        return {
```
Could this be an assignment to `collectibleInformation` instead? That would avoid having to encode the defaults in two places.

adonesky1 replied:
> That would avoid having to encode the defaults in two places.

Not sure I follow this? Where else are the defaults being encoded within this function?

adonesky1 replied:
Oh you mean the type is an encoding of the defaults

adonesky1 replied:
If we want to set it to the default version of the expected type for collectibleInformation - `ApiCollectible` - it would involve duplicating the encoded defaults again: 
```
 collectibleInformation = {
        name: null,
        description: null,
        asset_contract: {
          address: '',
          asset_contract_type: null,
          created_date: null,
          schema_name: null,
          symbol: null,
          total_supply: null,
          description: null,
          external_link: null,
          collection: {
            name: null,
            image_url: null,
          }
        },
        token_id: '',
        num_sales: null,
        background_color: null,
        image_url: null,
        image_preview_url: null,
        image_thumbnail_url: null,
        image_original_url: null,
        animation_url: null,
        animation_original_url: null,
        external_link: null,
        creator: {
          user: { username: '' },
          profile_img_url: '',
          address: '',
        },
        last_sale: {
          event_timestamp: '',
          total_price: '',
          transaction: { transaction_hash: '', block_hash: '' },
        },
      };
```

Gudahtt replied:
>Oh you mean the type is an encoding of the defaults

Sorry no, this is not what I meant. I don't understand the later comments in this thread either, maybe because of this misunderstanding.

All I'm saying is that the defaults are set in two places. Here, and below on lines 283-285. If this line is replaced with `collectibleInformation = {}`, it would be functionally equivalent but the defaults wouldn't need to be duplicated in two places.

adonesky1 replied:
The problem is that `collectibleInformation` is type `ApiCollectible` (line 239). So it can't be set to an empty object because that type is:

```
ApiCollectible {
  token_id: string;
  num_sales: number | null;
  background_color: string | null;
  image_url: string | null;
  image_preview_url: string | null;
  image_thumbnail_url: string | null;
  image_original_url: string | null;
  animation_url: string | null;
  animation_original_url: string | null;
  name: string | null;
  description: string | null;
  external_link: string | null;
  asset_contract: ApiCollectibleContract;
  creator: ApiCollectibleCreator;
  last_sale: ApiCollectibleLastSale | null;
}
```

adonesky1 replied:
`collectibleInformation` is type `ApiCollectible` while the return type for the function is `CollectibleMetadata`:

interface CollectibleMetadata {
  name: string | null;
  description: string | null;
  image: string | null;
  standard: string | null;
  favorite?: boolean;
  numberOfSales?: number;
  backgroundColor?: string;
  imagePreview?: string;
  imageThumbnail?: string;
  imageOriginal?: string;
  animation?: string;
  animationOriginal?: string;
  externalLink?: string;
  creator?: ApiCollectibleCreator;
  lastSale?: ApiCollectibleLastSale;
}

Gudahtt replied:
Why is that a problem? Can't you update the type of that variable to `CollectibleMetadata`? It must be assignable to that, given that that is what is being returned. I don't follow why this difference in types is a problem.

adonesky1 replied:
I might still be missing something about what you're suggesting, but the issue is that at the bottom of the function (after we've fetched data from either the proxy or directly from OpenSea) we take the shape of `ApiCollectible` (the expected return type from the API calls) and change it to the `CollectibleMetadata` shape which is how we want to store the data. In order re-type the `collectibleInformation` as `CollectibleMetadata` we would have to replace `handleFetch` with another function that combines the job of fetching and then (if successful) reshaping the data from `ApiCollectible` to `CollectibleMetadata` for each attempt.

Gudahtt replied:
Ok, I see what you mean. I won't block on this but I still think this function is unnecessarily complex, and that it would be better to set default values in one place rather than two. But at least they're encoded in the return type as well I guess.

adonesky1 replied:
I agree it could use a refactor! But perhaps we can leave that for another PR.

Gudahtt commented on src/assets/CollectiblesController.ts:
```diff
@@ -236,14 +237,27 @@ export class CollectiblesController extends BaseController<
+      logOrRethrowError(e);
+
+      if (this.openSeaApiKey) {
+        collectibleInformation = await handleFetch(
```
The "log and re-throw" pattern should be applied here as well. Otherwise the intended defaults will not be applied if both calls fail.

Maybe this "log and throw" step can be moved a function that is called in place of `handleFetch`?

adonesky1 replied:
@Gudahtt per my comment below I have sketched out two patterns to handle errors. Not sure which I prefer. When you have a moment could you provide input?

Gudahtt commented on src/assets/CollectiblesController.test.ts:
```diff
@@ -27,7 +27,7 @@ const MAINNET_PROVIDER = new HttpProvider(
 const SECOND_OWNER_ADDRESS = '0x500017171kasdfbou081';
 
-const OPEN_SEA_HOST = 'https://api.opensea.io';
+const OPEN_SEA_HOST_PROXY = 'https://proxy.metaswap.codefi.network/opensea/v1';
```
Could you add tests for the fallback behaviour as well?

adonesky1 replied:
Waiting to go back and complete tests til we have alignment on the rest.

adonesky1 replied:
added a test for the fallback behavior here: https://github.com/MetaMask/controllers/pull/805/commits/794069034080dfd7762402412f1d059204b1961a

adonesky1 commented:
@Gudahtt I've addressed some of your feedback, and fragmented the error handling approach into two slightly different patterns so you could visualize each and help me decide which one makes more sense. In `getCollectibleInformationFromApi` and `getCollectibleContractInformationFromApi` (`CollectiblesController.js`) I'm now using a modified fetch method (`fetchWithErrorHandling`) which catches errors and uses `logOrRethrowError`. This leads to a slightly strange pattern of having to check if what is returned is defined and then return that or move on in the function (to the retry or the default/null return value). In `getOwnerCollectibles` (`CollectiblesDetectionController.js`) I left the try/catches in place but added a nested try/catch around the retry attempt. It gets kindof hard to reason about when nested. Neither is pretty but they're the best options I could think of right now...

adonesky1 commented:
Ping here when you have time @Gudahtt 😅

Gudahtt commented on src/util.ts:
```diff
@@ -658,6 +658,28 @@ export async function handleFetch(request: string, options?: RequestInit) {
+  request: string,
+  options?: RequestInit,
+) {
+  let object;
```
Was this meant to be `response`? This is always `undefined` right now.

Gudahtt replied:
I don't follow why you didn't return on line 675. This variable doesn't seem needed at all.

adonesky1 replied:
yes on both counts... 🙄

Gudahtt commented on src/util.ts:
```diff
@@ -953,3 +975,20 @@ export function isTokenDetectionEnabledForNetwork(chainId: string): boolean {
+ */
+export function logOrRethrowError(error: any) {
+  if (
+    error.message?.match(/Fetch failed with status/u) ||
```
This would flag for any unsuccessful fetch. e.g. if we submitted malformed parameters to the API and got back a 4XX error, this would catch that case. But we'd want that case to bubble up instead so that we get a Sentry report explaining that our code is broken.

adonesky1 replied:
Ok. So do you think its okay to catch on `403`s in this method or perhaps pass an arg to `logOrRethrowError` with a platform specific error code we want to catch (`1020` in this case for instance)?

Gudahtt replied:
Coming back to the general advice given earlier, what we want is to ensure _expected errors_ are handled in an appropriate way for the situation, and that _unexpected errors_ are thrown so that we can learn about this unanticipated situation and determine the best way to handle it. Does that make sense?

Right now with this function, any non-OK response from the server is being treated as expected and logged to the console. You've correctly pointed out that we should anticipate the API key being unset or invalid some of the time (e.g. for development builds, or in case our key gets revoked). But I would not anticipate us getting any other 4xx response status codes apart from 403, not unless we made a mistake somewhere.

I don't know what 1020 is supposed to indicate, so I don't know whether we should encode 403 overall as being expected, or 1020 specifically. Maybe if they have API documentation guaranteeing that a missing or invalid API key always returns that code, we could encode that expectation. Otherwise, encoding that might risk this breaking in prod if they change the API to return a different code.

Gudahtt replied:
Probably best to just go by the 403. They confirmed they don't have docs for the 1020 error response, that it's just coming from cloudflare. So that might change in the future if they update how the API key check works or migrate to a new platform or something.

adonesky1 replied:
@Gudahtt I've pursued this approach here: https://github.com/MetaMask/controllers/pull/805/commits/9a0d739752b0028765e42340895610ee0d665886

Let me know what you think

adonesky1 replied:
@Gudahtt ping here when you have a moment

adonesky1 commented:
@Gudahtt adapted the helper method and applied to the `CollectibleDetectionController` case as well here: https://github.com/MetaMask/controllers/pull/805/commits/5115b4bc91f148eca9ddb86b79d798aa1d1d38ca. Still unclear on how to approach the error handling.

adonesky1 commented:
@sethkfman @gantunesr @wachunei if you have some time tomorrow to take a look at this, I'd really love to get this merged so we can start getting ready to release NFT's on the extension 😅

darkwing reviewed (dismissed):
Looks good; my only suggestion is making `https://proxy.metaswap.codefi.network/opensea/v1` a constant somewhere, as it's used multiple times.

adonesky1 commented:
> Looks good; my only suggestion is making https://proxy.metaswap.codefi.network/opensea/v1 a constant somewhere, as it's used multiple times.

consolidated these constants here: https://github.com/MetaMask/controllers/pull/805/commits/915c211c6d546c29de29fdfb9e1ab55cb95f5957

mcmire commented on src/assets/CollectiblesController.ts:
```diff
@@ -235,18 +249,40 @@ export class CollectiblesController extends BaseController<
-    } else {
-      collectibleInformation = await handleFetch(tokenURI);
+    // if an openSeaApiKey is set we should attempt to refetch calling directly to OpenSea
+    if (!collectibleInformation) {
+      if (this.openSeaApiKey) {
```
[nit] Could these if statements be combined?

adonesky1 replied:
yes!

adonesky1 replied:
done here: https://github.com/MetaMask/controllers/pull/805/commits/180b2802d9edef8ff9392b88ea3a3f76cc3373da

mcmire commented on src/assets/CollectiblesController.ts:
```diff
@@ -235,18 +249,40 @@ export class CollectiblesController extends BaseController<
+    // Attempt to fetch the data with the proxy
+    let collectibleInformation: ApiCollectible | undefined =
+      await fetchWithErrorHandling(
+        // the third arg defaults to true to use the proxy
```
If you have to leave a comment, then I'm wondering if this means that the third argument shouldn't have a default. What do you think?

adonesky1 replied:
made these positional with no default: https://github.com/MetaMask/controllers/pull/805/commits/180b2802d9edef8ff9392b88ea3a3f76cc3373da

mcmire commented on src/assets/CollectiblesController.ts:
```diff
@@ -235,18 +249,40 @@ export class CollectiblesController extends BaseController<
+    if (!collectibleInformation) {
+      if (this.openSeaApiKey) {
+        collectibleInformation = await fetchWithErrorHandling(
+          this.getCollectibleApi(contractAddress, tokenId, false),
```
When adding boolean arguments to functions/methods, it can be difficult to understand at first glance what those booleans signify. What do you think about having this method take an object instead of positional arguments?

adonesky1 replied:
Yeah I think that makes sense!

adonesky1 replied:
done here: https://github.com/MetaMask/controllers/pull/805/commits/180b2802d9edef8ff9392b88ea3a3f76cc3373da

mcmire commented on src/util.ts:
```diff
@@ -658,6 +658,42 @@ export async function handleFetch(request: string, options?: RequestInit) {
+export async function fetchWithErrorHandling(
+  request: string,
+  options?: RequestInit,
+  timeout?: number | null,
```
Would `number | undefined` make more sense here? I feel like `null` usually implies that `null` is intentional, whereas `undefined` usually implies that it's left out.

adonesky1 replied:
fixed here: https://github.com/MetaMask/controllers/pull/805/commits/180b2802d9edef8ff9392b88ea3a3f76cc3373da

mcmire commented on src/util.ts:
```diff
@@ -658,6 +658,42 @@ export async function handleFetch(request: string, options?: RequestInit) {
+  let result;
+  try {
+    if (timeout) {
+      result = Promise.race([
```
Wouldn't this be the `response`? It seems like the logic is slightly different between `timeout` and no `timeout`, since `result = await response.json()` is not being called in this case.

adonesky1 replied:
Good catch, both of these cases should use `handleFetch` which just returns `response.json()`.

mcmire commented on src/util.ts:
```diff
@@ -953,3 +989,27 @@ export function isTokenDetectionEnabledForNetwork(chainId: string): boolean {
+ * @param error - Caught error that we should either rethrow or log to console.
+ * @param codesToCatch - array of error codes for errors we want to catch and log in a particular context
+ */
+export function logOrRethrowError(error: any, codesToCatch?: number[]) {
```
`any` implies that `error` can be `null` or `undefined`. Should we check `error` for that?

adonesky1 replied:
I believe this can/should be typeof `Error`

adonesky1 replied:
done here: https://github.com/MetaMask/controllers/pull/805/commits/d29d97cd7a86c5bd82e20ac2d73a0f34ff43171c

sethkfman commented on src/assets/CollectiblesController.ts:
```diff
@@ -170,24 +174,34 @@ export class CollectiblesController extends BaseController<
-        return `https://api.opensea.io/api/v1/asset_contract/${contractAddress}`;
+
+    if (chainId === RINKEBY_CHAIN_ID) {
+      return `https://testnets-api.opensea.io/api/v1/asset_contract/${contractAddress}`;
```
I would replace string with constant `OPENSEA_TEST_API_URL`

adonesky1 replied:
done here: https://github.com/MetaMask/controllers/pull/805/commits/180b2802d9edef8ff9392b88ea3a3f76cc3373da

mcmire commented on src/util.ts:
```diff
@@ -953,3 +989,27 @@ export function isTokenDetectionEnabledForNetwork(chainId: string): boolean {
+ * @param error - Caught error that we should either rethrow or log to console.
+ * @param codesToCatch - array of error codes for errors we want to catch and log in a particular context
+ */
+export function logOrRethrowError(error: any, codesToCatch?: number[]) {
+  const includesErrorCodeToCatch = codesToCatch?.length
+    ? codesToCatch?.some((code) =>
+        error.message?.includes(`Fetch failed with status '${code}'`),
+      )
+    : false;
```
`some` already returns a boolean (and returns false if the array is empty). Also, instead of having to check for `codesToCatch`, what do you think about defaulting it to an empty array? So this would be:

``` suggestion
export function logOrRethrowError(error: any, codesToCatch?: number[] = []) {
  const includesErrorCodeToCatch = codesToCatch.some((code) =>
    error.message?.includes(`Fetch failed with status '${code}'`),
  );
```

adonesky1 replied:
done here: https://github.com/MetaMask/controllers/pull/805/commits/180b2802d9edef8ff9392b88ea3a3f76cc3373da

adonesky1 commented on src/util.ts:
```diff
@@ -658,6 +658,42 @@ export async function handleFetch(request: string, options?: RequestInit) {
+ * @returns The fetch response JSON data or undefined (if error occurs).
+ */
+export async function fetchWithErrorHandling(
+  request: string,
```
Yep sounds right!

adonesky1 commented on src/assets/CollectibleDetectionController.ts:
```diff
@@ -312,72 +329,71 @@ export class CollectibleDetectionController extends BaseController<
+            last_sale && { lastSale: last_sale },
+          );
+
+          await this.addCollectible(address, token_id, collectibleMetadata, {
```
This PR actually doesn't change this behavior, it simply removes the `safelyExecute` wrapper. But `this.addCollectible` does not affect `ignoredCollectibles` so I believe we can safely access `ignoredCollectibles` and write to state with `addCollectibles()` in parallel.

mcmire replied:
Okay cool.

mcmire reviewed (commented):
One more thing I noted above ^

adonesky1 commented:
@mcmire @sethkfman ready for another look when you have a chance 😄

adonesky1 commented on src/assets/CollectibleDetectionController.test.ts:
```diff
@@ -19,6 +18,9 @@ describe('CollectibleDetectionController', () => {
   let assetsContract: AssetsContractController;
 
+  const getOpenSeaApiKeyStub = jest.fn();
+  nock.disableNetConnect();
```
Yup was causing havoc! Added here: https://github.com/MetaMask/controllers/pull/805/commits/002e08b6cb95e22acc8860e02789f61f212db895

adonesky1 commented on src/util.ts:
```diff
@@ -953,3 +994,30 @@ export function isTokenDetectionEnabledForNetwork(chainId: string): boolean {
+    error instanceof Error &&
+    (includesErrorCodeToCatch ||
+      error.message?.includes('Failed to fetch') ||
+      error.message === 'timeout')
```
done here: https://github.com/MetaMask/controllers/pull/805/commits/1a0b10f1d5f1b9dc628b1bfca0ab995500eed778

Gudahtt commented:
Doesn't need to be blocked on my review - everything I commented on has been resolved
